src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs

```diff
@@ -393,6 +413,60 @@ private void CreateCountPath(ODataPath currentPath, OpenApiConvertSettings conve
             AppendPath(countPath);
         }
 
+        /// <summary>
+        /// Create OData type cast paths.
+        /// </summary>
+        /// <param name="currentPath">The current OData path.</param>
+        /// <param name="convertSettings">The settings for the current conversion.</param>
+        /// <param name="structuredType">The type that is being inherited from to which this method will add downcast path segments.</param>
+        /// <param name="annotable">The annotable navigation source to read cast annotations from.</param>
+        /// <param name="targetsMany">Whether the annotable navigation source targets many entities.</param>
+        private void CreateTypeCastPaths(ODataPath currentPath, OpenApiConvertSettings convertSettings, IEdmStructuredType structuredType, IEdmVocabularyAnnotatable annotable, bool targetsMany)
+        {
+            if(currentPath == null) throw new ArgumentNullException(nameof(currentPath));
+            if(convertSettings == null) throw new ArgumentNullException(nameof(convertSettings));
+            if(structuredType == null) throw new ArgumentNullException(nameof(structuredType));
+            if(annotable == null) throw new ArgumentNullException(nameof(annotable));
+            if(!convertSettings.EnableODataTypeCast) return;
+
+            var annotedTypeNames = GetDerivedTypeConstaintTypeNames(annotable);
+            
+            if(!annotedTypeNames.Any() && convertSettings.RequireDerivedTypesConstraintForODataTypeCastSegments) return; // we don't want to generate any downcast path item if there is no type cast annotation.
```

Comment:
don't need to check the constaint items? 
I mean to check whether the constraint can apply to a certain type cast?

Reply:
the the consumer requires the annotation before adding any type cast segment, and there's no annotation present, there's no point continuing further